### PR TITLE
[XLA:CPU] Fix out-of-bounds read and crash in spatial convolution with negative padding

### LINKED ISSUE #48713: [XLA:CPU] Wrong results / out-of-bounds read in f32 convolution with negative low padding
A `conv_general_dilated` on f32 whose spatial dimension has **negative low padding** and a
positive high padding of at most `|low|` reads outside the input buffer: the output positions that
should be zero padding contain data from the next row, and past the end of the buffer, uninitialised
memory. The identical convolution in `int32` (which goes through the elemental emitter rather than
Eigen) is correct, as is the `>16`-channel case that takes the YNNPACK path.

### Reproducer

```python
import os
os.environ["JAX_PLATFORMS"] = "cpu"
import numpy as np, jax, jax.numpy as jnp
from jax import lax

H, W, C = 15, 6, 8
x = (100 * np.arange(H)[:, None, None] + 10 * np.arange(W)[None, :, None] + np.arange(C)[None, None, :])
x = x.reshape(1, H, W, C)                                  # x[0,h,w,c] = 100h + 10w + c  (NHWC)
k = np.eye(C).reshape(1, 1, C, C)                          # 1x1 identity kernel (HWIO): out[...,c] = padded_x[...,c]
dn = lax.ConvDimensionNumbers((0, 3, 1, 2), (3, 2, 0, 1), (0, 3, 1, 2))
pads = ((0, 0), (-5, 1))                                   # W: crop 5 -> [x5], then pad one zero -> [x5, 0]

for dtype in (jnp.float32, jnp.int32):
    f = lambda a, b: lax.conv_general_dilated(a, b, (1, 1), pads, dimension_numbers=dn)
    xa, ka = jnp.asarray(x, dtype), jnp.asarray(k, dtype)
    expected = np.stack([x[0, :, 5, :], np.zeros((H, C))], axis=1)[None]      # (1, 15, 2, 8)
    eager = np.asarray(f(xa, ka)); jitted = np.asarray(jax.jit(f)(xa, ka))
    print(dtype.__name__, "row0 expected:", expected[0, 0].tolist())
    print(dtype.__name__, "row0 eager   :", eager[0, 0].tolist(), "WRONG" if not np.array_equal(eager, expected) else "ok")
    print(dtype.__name__, "row0 jit     :", jitted[0, 0].tolist(), "WRONG" if not np.array_equal(jitted, expected) else "ok")
    print(dtype.__name__, "last row jit  :", jitted[0, -1, 1].tolist(), "(expected zeros)")
```

Output on my machine:

```
float32 row0 expected: [[50.0, ..., 57.0], [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0]]
float32 row0 eager   : [[50.0, ..., 57.0], [100.0, 101.0, 102.0, 103.0, 104.0, 105.0, 106.0, 107.0]] WRONG
float32 row0 jit     : [[50.0, ..., 57.0], [100.0, 101.0, 102.0, 103.0, 104.0, 105.0, 106.0, 107.0]] WRONG
float32 last row jit : [0.0, 0.0, 0.0, 0.0, -4891756.0, 0.0, 0.0, 0.0]   (expected zeros)
int32   row0 eager   : [[50, ..., 57], [0, 0, 0, 0, 0, 0, 0, 0]] ok
int32   row0 jit     : [[50, ..., 57], [0, 0, 0, 0, 0, 0, 0, 0]] ok
```

### Analysis

With a 1x1 identity kernel the output is just the padded input, so the padded `W` axis must be
`[x[:, 5, :], 0]`. Instead the second position holds `x[0, h+1, 0, :]`, and in the last row it holds
whatever follows the buffer (`-4891756.0` here; other runs gave `-3.7e27`, `4.9e36`).

The suspect is the `hasPadding()` helper used by the Eigen image-patch path
(`xla/tsl/framework/convolution/eigen_spatial_convolutions-inl.h`), which only considers padding
*positive*, so a negative before-padding leaves the "no padding" fast path selected and the patch
extraction indexes the input directly.

Affected padding pairs I checked on the `W` axis: `(-6,1) (-6,2) (-5,1) (-5,2) (-4,3) (-3,1) (-7,3)
(-8,3)`, and on `H`: `(-15,1) (-15,2) (-14,2)`. A positive low padding with a negative high padding
(`(1,-6)`, `(2,-7)`) is fine. Negative convolution padding is a documented feature and JAX emits it
in convolution VJPs, so this is reachable from ordinary gradient code.

### Environment

- `jax` 0.11.1, `jaxlib` 0.11.1 (pip), CPU backend, `JAX_PLATFORMS=cpu`
- Linux x86-64 (AVX-512), Python 3.12

### Notes

Found by a differential fuzzer comparing XLA:CPU against a NumPy oracle written from the operation semantics (validated by hand and against `jax._src.lax_reference`), then minimised. I am filing this as an issue rather than a pull request, per the repository's contribution guidelines: this report was prepared with the help of an AI coding assistant (Claude), and I reproduced every result above myself on the environment listed.

### PR DESCRIPTION
**Pull request by jdymitarai**
#ai-generated

### Root Cause
In `xla/tsl/framework/convolution/eigen_spatial_convolutions-inl.h`, `hasPadding()` only verified positive padding (`m_rowPaddingTop > 0`, `m_colPaddingLeft > 0`) and omitted padding offsets from `last_row` and `last_col` calculations. When a spatial convolution specifies negative low padding (cropping) combined with high padding, or crops an entire dimension, `hasPadding()` incorrectly returned `false`, selecting the unpadded fast path (`packetNoPadding`). This resulted in out-of-bounds reads into uninitialized memory and `SIGSEGV` / access violations when vectorizing across wide channel dimensions.

### Fix Approach
Update `hasPadding()` to:
1. Include `-m_rowPaddingTop` and `-m_colPaddingLeft` in `first_row`/`first_col` and `last_row`/`last_col` calculations.
2. Return `true` whenever the first or last accessed spatial coordinates fall outside `[0, input_dim)`.
3. Preserve the vectorized fast path when cropping strictly within input bounds.

### Test Verification
- Added `NegativeLowPaddingSpatialConvolution` and `CropEntireDimensionSpatialConvolution` unit tests in `eigen_spatial_convolutions_test.cc`.
- Verified against the JAX reproducer in #48713, confirming padded boundaries evaluate to zero without out-of-bounds reads or crashes.

Fixes #48713

### CODE REVIEW

File: xla/tsl/framework/convolution/eigen_spatial_convolutions-inl.h [hunk 1]
```diff
@@ -881,19 +881,27 @@ class TensorContractionSubMapper<
...
-    // Non zero padding before.
-    if (m_base_mapper.m_rowPaddingTop > 0) return true;
-    if (m_base_mapper.m_colPaddingLeft > 0) return true;
+    if (m_base_mapper.m_outputRows <= 0 || m_base_mapper.m_outputCols <= 0) {
+      return false;
+    }
+
+    // Check row bounds.
+    const Index first_row = -m_base_mapper.m_rowPaddingTop;
+    if (first_row < 0 || first_row >= m_base_mapper.m_inputRows) return true;
```
Comment (akuegel) [inline xla/tsl/framework/convolution/eigen_spatial_convolutions-inl.h:890]: nit: Please do not put the `return true` on the same line. We prefer {} block after if statements.

File: xla/tsl/framework/convolution/eigen_spatial_convolutions-inl.h [hunk 2]
```diff
@@ -881,19 +881,27 @@ class TensorContractionSubMapper<
...
 
-    // Non zero padding after in rows.
     const Index last_row =
-        (m_base_mapper.m_outputRows - 1) * m_base_mapper.m_row_strides;
-    if (last_row + (patchRows() - 1) >= m_base_mapper.m_inputRows) return true;
+        (m_base_mapper.m_outputRows - 1) * m_base_mapper.m_row_strides -
+        m_base_mapper.m_rowPaddingTop + (patchRows() - 1);
+    if (last_row < 0 || last_row >= m_base_mapper.m_inputRows) return true;
```
Comment (akuegel) [inline xla/tsl/framework/convolution/eigen_spatial_convolutions-inl.h:895]: Same here.

File: xla/tsl/framework/convolution/eigen_spatial_convolutions-inl.h [hunk 3]
```diff
@@ -881,19 +881,27 @@ class TensorContractionSubMapper<
...
+
+    // Check col bounds.
+    const Index first_col = -m_base_mapper.m_colPaddingLeft;
+    if (first_col < 0 || first_col >= m_base_mapper.m_inputCols) return true;
```
Comment (akuegel) [inline xla/tsl/framework/convolution/eigen_spatial_convolutions-inl.h:899]: Same here.

File: xla/tsl/framework/convolution/eigen_spatial_convolutions-inl.h [hunk 4]
```diff
@@ -881,19 +881,27 @@ class TensorContractionSubMapper<
...
 
-    // Non zero padding after in cols.
     const Index last_col =
-        (m_base_mapper.m_outputCols - 1) * m_base_mapper.m_col_strides;
-    if (last_col + (patchCols() - 1) >= m_base_mapper.m_inputCols) return true;
+        (m_base_mapper.m_outputCols - 1) * m_base_mapper.m_col_strides -
+        m_base_mapper.m_colPaddingLeft + (patchCols() - 1);
+    if (last_col < 0 || last_col >= m_base_mapper.m_inputCols) return true;
```
Comment (akuegel) [inline xla/tsl/framework/convolution/eigen_spatial_convolutions-inl.h:904]: Same here.

**Comment by akuegel** [comment]
@jdymitarai did you have a chance to work on the small changes I requested?
